django_keycloak_auth/middleware.py

```diff
@@ -18,79 +18,92 @@
-            
-        if self.config['KEYCLOAK_CLIENT_SECRET_KEY'] is None:
-            raise Exception("The mandatory KEYCLOAK_CLIENT_SECRET_KEY configuration variables has not defined.")
+            raise ValueError("The mandatory KEYCLOAK configuration variables has not defined.")
+
+        if config['KEYCLOAK_SERVER_URL'] is None:
+            raise ValueError("The mandatory KEYCLOAK_SERVER_URL configuration variables has not defined.")
+
+        if config['KEYCLOAK_REALM'] is None:
+            raise ValueError("The mandatory KEYCLOAK_REALM configuration variables has not defined.")
+
+        if config['KEYCLOAK_CLIENT_ID'] is None:
+            raise ValueError("The mandatory KEYCLOAK_CLIENT_ID configuration variables has not defined.")
+
+        if config['KEYCLOAK_CLIENT_SECRET_KEY'] is None:
+            raise ValueError("The mandatory KEYCLOAK_CLIENT_SECRET_KEY configuration variables has not defined.")
+
+
+class KeycloakMiddleware:
+
+    def __init__(self, get_response):
+        
+        # Read Keycloak configurations
+        self.keycloak_config = KeycloakConfig()
+
+        # Django response
+        self.get_response = get_response
 
         # Create Keycloak instance
-        self.keycloak = KeycloakConnect(server_url=self.server_url,
-                                        realm_name=self.realm,
-                                        client_id=self.client_id,
-                                        client_secret_key=self.client_secret_key)
+        self.keycloak = KeycloakConnect(server_url=self.keycloak_config.server_url,
+                                        realm_name=self.keycloak_config.realm,
+                                        client_id=self.keycloak_config.client_id,
+                                        client_secret_key=self.keycloak_config.client_secret_key)
 
     def __call__(self, request):
-        return self.get_response(request)
+        return self.get_response(request)      
 
     def process_view(self, request, view_func, view_args, view_kwargs):
-
-        # for now there is no role assigned yet
+        
+        # for now there is no role assigned yet and no userinfo defined
         request.roles = []
+        request.userinfo = []
 
         # Checks the URIs (paths) that doesn't needs authentication        
         if hasattr(settings, 'KEYCLOAK_EXEMPT_URIS'):
             path = request.path_info.lstrip('/')
             if any(re.match(m, path) for m in settings.KEYCLOAK_EXEMPT_URIS):
                 return None
-        
-        # Read if View has attribute 'keycloak_roles'
+
+        # There's condictions for these view_func.cls:
+        # 1) @api_view -> view_func.cls is WrappedAPIView (validates in 'keycloak_roles' in decorators.py) -> True
+        # 2) When it is a APIView, ViewSet or ModelViewSet with 'keycloak_roles' attribute -> False
+        is_api_view = True if str(view_func.cls) == "<class 'core.views.WrappedAPIView'>" else False
```

Comment:
thanks for the explanation! indeed, I had not looked at the api_view source code before. Now I think I've figured out why I'm getting a 403 error. here, the name of the type. for me it's <class 'my_project.my_views_folder.my_view_file.WrappedAPIView'>

Reply:
Super cool! Good idea. It makes it simpler.
I've already changed it

Thanks!

Reply:
everything works now, thank you! 🙂